### PR TITLE
Virtio exp

### PR DESCRIPTION
**Pull request by danmilea**
virtio-mmio framework for lib open-amp

### CODE REVIEW

File: lib/include/openamp/virtio.h [hunk 1]
```diff
@@ -24,12 +24,14 @@ extern "C" {
...
 #define VIRTIO_ID_SCSI       0x08UL
 #define VIRTIO_ID_9P         0x09UL
+#define VIRTIO_ID_SCMI       0x20UL
 #define VIRTIO_DEV_ANY_ID    (-1)UL
 
 /* Status byte for guest to report progress. */
 #define VIRTIO_CONFIG_STATUS_ACK       0x01
 #define VIRTIO_CONFIG_STATUS_DRIVER    0x02
 #define VIRTIO_CONFIG_STATUS_DRIVER_OK 0x04
+#define VIRTIO_CONFIG_FEATURES_OK 0x08
```
Comment (xiaoxiang781216) [inline lib/include/openamp/virtio.h:34]: align 0x08 with other line

File: lib/include/openamp/virtqueue.h [hunk 2]
```diff
@@ -189,6 +189,10 @@ void *virtqueue_get_buffer(struct virtqueue *vq, uint32_t *len, uint16_t *idx);
 
 void *virtqueue_get_available_buffer(struct virtqueue *vq, uint16_t *avail_idx,
 				     uint32_t *len);
+uint16_t virtqueue_buffer_writable(struct virtqueue *vq, uint16_t avail_idx);
+
+void *virtqueue_get_next_buffer(struct virtqueue *vq, uint16_t *avail_idx,
+				     uint32_t *len);
```
Comment (xiaoxiang781216) [inline lib/include/openamp/virtqueue.h:196]: aglin uint32_t to struct at line 194

File: lib/virtio/virtqueue.c [hunk 3]
```diff
@@ -287,6 +289,36 @@ void *virtqueue_get_available_buffer(struct virtqueue *vq, uint16_t *avail_idx,
 	return buffer;
 }
 
+uint16_t virtqueue_buffer_writable(struct virtqueue *vq, uint16_t avail_idx)
```
Comment (xiaoxiang781216) [inline lib/virtio/virtqueue.c:292]: should we use bool instead uint16_t

File: lib/include/openamp/virtio.h [hunk 4]
```diff
@@ -91,6 +92,12 @@ struct virtio_device_id {
 #define VIRTIO_TRANSPORT_F_START      28
 #define VIRTIO_TRANSPORT_F_END        32
 
+#define VIRTIO_F_ACCESS_PLATFORM	33
```
Comment (xiaoxiang781216) [inline lib/include/openamp/virtio.h:95]: align 33 with 32 at line 93

Comment (xiaoxiang781216) [inline lib/include/openamp/virtio.h:95]: should we use VIRTIO_ACCESS_F_PLATFORM

File: lib/include/openamp/virtio.h [hunk 5]
```diff
@@ -91,6 +92,12 @@ struct virtio_device_id {
...
+
+/* packed virtqueue layout. */
+#define VIRTIO_F_RING_PACKED		34
```
Comment (xiaoxiang781216) [inline lib/include/openamp/virtio.h:98]: same comment as VIRTIO_F_ACCESS_PLATFORM

File: lib/virtio_mmio/virtio_mmio_queue.c [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+}
+#endif
+
+void *virtqueue_dequeue(struct virtqueue *vq, uint32_t *len)
+{
+    return virtqueue_get_buffer(vq, len, NULL);
+}
+
+#if 0
+void virtqueue_init(const struct device *dev, unsigned int idx, struct virtqueue *vq, void (*cb)(void*), void *cb_arg)
```
Comment (xiaoxiang781216) [inline lib/virtio_mmio/virtio_mmio_queue.c:66]: remove the unused code

File: lib/virtio_mmio/virtio_mmio_queue.c [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    return virtqueue_add_buffer(vq, vb, writable ? 0 : 1, writable, cookie);
+}
+
+int virtqueue_enqueue(struct virtqueue *vq, void *cookie, struct iovec *iov, size_t readable, size_t writable)
+{
+    return virtqueue_add_buffer(vq, (struct virtqueue_buf *)iov, readable, writable, cookie);
+}
+
+#if 0
+void *virtqueue_dequeue_buf(struct virtqueue *vq, uint32_t *len)
```
Comment (xiaoxiang781216) [inline lib/virtio_mmio/virtio_mmio_queue.c:29]: remove too

File: lib/include/openamp/virtio_config.h [hunk 8]
```diff
@@ -0,0 +1,155 @@
+/*
+ * Virtio platform device driver
```
Comment (xiaoxiang781216) [inline lib/include/openamp/virtio_config.h:2]: why name the file virtio_config.h, but all macros are related to virtio_mmio.h

**Review by arnopo — COMMENTED** [review]
Hi @danmilea 

Quite a lot of comment... I reviewed considering that i could be applied on main branch.

That said  this branch has been created first as a experimental branch, to initiate virtio evolution based on your demo and @vingu-linaro work.
So I could integrate it like this for a demo need.

Now It depends how we would like to proceed for a collaborative work.
It will be necessary at a time to clean the code and make it compliant with the checkpatch...

File: lib/include/openamp/virtio.h [hunk 9]
```diff
@@ -79,6 +81,7 @@ struct virtio_device_id {
  */
 #define VIRTIO_F_BAD_FEATURE (1 << 30)
 
+#define VIRTIO_F_VERSION_1 32
```
Comment (arnopo) [inline lib/include/openamp/virtio.h:85]: to move after VIRTIO_TRANSPORT_F_END declaration

File: lib/include/openamp/virtio.h [hunk 10]
```diff
@@ -24,12 +24,14 @@ extern "C" {
 #define VIRTIO_ID_RPMSG      0x07UL /* remote processor messaging */
 #define VIRTIO_ID_SCSI       0x08UL
 #define VIRTIO_ID_9P         0x09UL
+#define VIRTIO_ID_SCMI       0x20UL
```
[the remark this answers is not in the document]
Comment (danmilea) [reply lib/include/openamp/virtio.h:27]: The actual ID in decimal is 32, which matches the SCMI device ID in https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.html#x1-2160005

File: scripts/checkpatch.pl [hunk 11]
```diff
@@ -3054,8 +3054,9 @@ sub process {
 				    $newindent ne $goodspaceindent) {
 
 					if (CHK("PARENTHESIS_ALIGNMENT",
-						"Alignment should match open parenthesis\n" . $hereprev) &&
+						"Alignment should match open parenthesis newindent = " . "$newindent" . "gt =  " . "$goodtabindent" . "gs = " . "$goodspaceindent" . "\n" . $hereprev) &&
```
Comment (arnopo) [inline scripts/checkpatch.pl]: to add checkpatch fix in a separate patch (also on main branch if it fix an issue)

File: lib/include/openamp/virtio_mmio.h [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Control registers */
+
+/* Magic value ("virt" string) - Read Only */
+#define VIRTIO_MMIO_MAGIC_VALUE		0x000
+
+#define MAGIC_VALUE ('v' | ('i' << 8) | ('r' << 16) | ('t' << 24))
```
Comment (arnopo) [inline lib/include/openamp/virtio_mmio.h:24]: definitions, variables, structures and function that are exported should be prefixed by `virtio_mmio`

File: lib/include/openamp/virtio_mmio.h [hunk 13]
```diff
@@ -0,0 +1,485 @@
...
+ * @param[in] dev Pointer to device structure.
+ * @param[in] features Features supported by the driver as a bitfield.
+ *
+ * @return N/A.
+ */
+
+inline void virtio_set_features(struct virtio_device *vdev, uint32_t features)
+{
+	return vdev->func->set_features(vdev, features);
+}
```
Comment (arnopo) [inline lib/include/openamp/virtio_mmio.h:334]: these functions above related to the virtio_device struct could be moved in virtio.h ( and some other below)

File: lib/include/openamp/virtio_mmio.h [hunk 14]
```diff
@@ -0,0 +1,485 @@
...
+					      void *cb_arg,
+					      const char *vq_name);
+
+/**
+ * @brief Notify device of virtqueue activity.
+ * @param[in] vq Pointer to virtqueue structure.
+ * @return N/A.
+ */
+
+inline void virtqueue_notify(struct virtqueue *vq)
```
Comment (arnopo) [inline lib/include/openamp/virtio_mmio.h:373]: what about using `virtqueue_kick` instead?

File: lib/virtio_mmio/virtio_mmio.c [hunk 15]
```diff
@@ -0,0 +1,416 @@
...
+}
+
+static void virtio_mmio_reset_device(struct virtio_device *vdev)
+{
+	virtio_mmio_write32(vdev, VIRTIO_MMIO_STATUS, 0);
+}
+
+static void virtio_mmio_notify(struct virtqueue *vq)
+{
+	virtio_mmio_write32(vq->vq_dev, VIRTIO_MMIO_QUEUE_NOTIFY, vq->vq_queue_index);
```
Comment (arnopo) [inline lib/virtio_mmio/virtio_mmio.c:125]: Add a comment here specifying that the VIRTIO_F_NOTIFICATION_DATA feature is not supported.
could be nice to return an error also in _virtio_mmio_set_features

File: lib/include/openamp/virtio_mmio.h [hunk 16]
```diff
@@ -0,0 +1,485 @@
...
+ *
+ * @param[in] vq Pointer to virtqueue structure.
+ * @param[out] len Pointer to length of transferred data. Can be NULL.
+ *
+ * @return cookie parameter to virtqueue_enqueue_buf()/virtqueue_enqueue()
+ * if a buffer is available.
+ * @return NULL if there is no buffer available.
+ */
+
+void *virtqueue_dequeue(struct virtqueue *vq, uint32_t *len);
```
Comment (arnopo) [inline lib/include/openamp/virtio_mmio.h:434]: I waiting to see how these 3 functions will evolve... waiting that could you create a virtio_mmio_queue.h for them?

File: lib/include/openamp/virtio_mmio.h [hunk 17]
```diff
@@ -0,0 +1,485 @@
...
+ *
+ * @return 1 if virtqueue is full.
+ * @return 0 if virtqueue is not full.
+ */
+
+inline int virtqueue_full(struct virtqueue *vq)
+{
+	return (vq->vq_free_cnt == 0);
+}
+
```
Comment (arnopo) [inline lib/include/openamp/virtio_mmio.h:463]: The 2 functions above could be moved in virtqueue.h

**Review by arnopo — COMMENTED** [review]
Few new comments for my second round

File: lib/virtio/virtqueue.c [hunk 18]
```diff
@@ -287,6 +287,41 @@ void *virtqueue_get_available_buffer(struct virtqueue *vq, uint16_t *avail_idx,
...
+	uint16_t flags;
+	void *buffer;
+
+	VQUEUE_BUSY(vq);
+
+	VRING_INVALIDATE(vq->vq_ring.desc[*avail_idx]);
+	flags = vq->vq_ring.desc[*avail_idx].flags;
+
+	if (!(flags & VRING_DESC_F_NEXT))
+		return NULL;
```
Comment (arnopo) [inline lib/virtio/virtqueue.c:312]: `VQUEUE_IDLE(vq); ` missing before return

**Review by arnopo — CHANGES_REQUESTED** [review]
A potential deadlock to fix. Then ok for me to merge it in the virtio_exp branch
@edmooring : do you want to review it before I merge?  At the end a new round of review will be performed when merging in the main branch...

File: lib/virtio/virtqueue.c [hunk 19]
```diff
@@ -287,6 +287,41 @@ void *virtqueue_get_available_buffer(struct virtqueue *vq, uint16_t *avail_idx,
...
+}
+
+void *virtqueue_get_next_buffer(struct virtqueue *vq, uint16_t *avail_idx,
+				uint32_t *len)
+{
+	uint16_t flags;
+	void *buffer;
+
+	VRING_INVALIDATE(vq->vq_ring.desc[*avail_idx]);
+	flags = vq->vq_ring.desc[*avail_idx].flags;
```
Comment (arnopo) [inline lib/virtio/virtqueue.c:307]: As we access to the vring here i would prefer to to keep `VQUEUE_BUSY` protection... you can use a goto if you want to have a single call to  `VQUEUE_IDLE()` in  he function

**Review by arnopo — APPROVED** [review]

**Pull request merged**